chain/consensus/mir/conf_manager.go

```diff
@@ -0,0 +1,270 @@
+	NextConfigurationNumberKey = datastore.NewKey("mir/next-config-number")
+	// NextAppliedConfigurationNumberKey is used to store AppliedConfigurationNumber
+	// that is the maximum configuration request number that has been applied.
+	NextAppliedConfigurationNumberKey = datastore.NewKey("mir/next-applied-config-number")
+	// ReconfigurationVotesKey is used to store configuration votes.
+	ReconfigurationVotesKey = datastore.NewKey("mir/reconfiguration-votes")
+)
+
+var _ client.Client = &ConfigurationManager{}
+
+type ConfigurationManager struct {
+	ctx           context.Context // Parent context
+	ds            db.DB           // Persistent storage.
+	id            string          // Parent ID.
+	nextReqNo     uint64          // Configuration number.
+	nextAppliedNo uint64          // Last applied number.
+}
+
+func NewConfigurationManager(ctx context.Context, ds db.DB, id string) (*ConfigurationManager, error) {
+	cm := &ConfigurationManager{
+		ctx:           ctx,
+		ds:            ds,
+		id:            id,
+		nextReqNo:     0,
+		nextAppliedNo: 0,
+	}
+	err := cm.recover()
+	if err != nil {
+		return nil, err
+	}
+	return cm, nil
+}
+
+func (cm *ConfigurationManager) NewTX(_ uint64, data []byte) (*mirproto.Request, error) {
+	r := mirproto.Request{
+		ClientId: cm.id,
+		ReqNo:    cm.nextReqNo,
+		Type:     ConfigurationRequest,
+		Data:     data,
+	}
+
+	if err := cm.storeRequest(&r, cm.nextReqNo); err != nil {
+		log.With("validator", cm.id).Errorf("unable to store configuration request: %v", err)
+		return nil, err
+	}
+
+	// If a request with number n was stored then the stored configuration nonce can be no more than n+1.
+	// That is possible if a node crashes here.
+
+	cm.nextReqNo++
+	cm.storeNextConfigurationNumber(cm.nextReqNo)
+
+	return &r, nil
+}
+func (cm *ConfigurationManager) Done(txNo t.ReqNo) error {
+	cm.nextAppliedNo = uint64(txNo) + 1
+	cm.storeNextAppliedConfigurationNumber(cm.nextAppliedNo)
+	cm.removeRequest(uint64(txNo))
+	return nil
+}
+
+func (cm *ConfigurationManager) Pending() (reqs []*mirproto.Request, err error) {
+	for i := cm.nextAppliedNo; i < cm.nextReqNo; i++ {
+		r, err := cm.getRequest(i)
+		if err != nil {
+			return nil, err
+		}
+		reqs = append(reqs, r)
+	}
+	return
+}
+
+func (cm *ConfigurationManager) Sync() error {
+	return fmt.Errorf("not implemented")
+}
+
+// recover function recovers configuration number, and configuration requests that may be not applied.
+func (cm *ConfigurationManager) recover() error {
+	nextReqNo := cm.getNextConfigurationNumber()
+	appliedNumber := cm.getAppliedConfigurationNumber()
+
+	if nextReqNo == appliedNumber && nextReqNo == 0 {
+		cm.nextReqNo = 0
+		cm.nextAppliedNo = 0
+		return nil
+	}
+	if appliedNumber > nextReqNo {
+		return fmt.Errorf("validator %v has incorrect configuration numbers: %d, %d", cm.id, appliedNumber, nextReqNo)
+	}
+
+	cm.nextAppliedNo = appliedNumber
+	cm.nextReqNo = nextReqNo
+
+	_, err := cm.getRequest(nextReqNo + 1)
```

Comment:
I guess that here you are determining if the next configuration was stored but not apply, or if we are starting from scratch, right? Can you add a brief description of what you are trying to achieve with this function?

Reply:
Fixed.